### PR TITLE
[SOL-1628] "N" times use coupon code

### PR DESCRIPTION
**Pull request by vkaracic**
This PR adds an ability to create coupons with a set usage number for `ONCE_PER_CUSTOMER` limitation type vouchers. 
A new field appears on the coupon creation form when the user selects `ONCE_PER_CUSTOMER` which sets the `max_global_applications` for the offer. Also new columns are added to the coupon report which adds new information about the usage count and order.

UI screenshot:
screenshot from 2016-04-04 09-00-22

Example coupon report document:
https://drive.google.com/a/edx.org/file/d/0ByBQUYJ4LMyAMWlQWWpWSWozQTg/view?usp=sharing

https://openedx.atlassian.net/browse/SOL-1628

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=652

File: ecommerce/extensions/api/serializers.py [hunk 1]
```diff
@@ -433,11 +434,19 @@ def get_note(self, obj):
         except AttributeError:
             return None
 
+    def get_usage_num(self, obj):
```
Comment (mjfrey) [inline ecommerce/extensions/api/serializers.py]: can we use the name "usage_count"  everywhere rather than usage_num

File: ecommerce/extensions/api/serializers.py [hunk 2]
```diff
@@ -400,6 +400,7 @@ class CouponSerializer(ProductPaymentInfoMixin, serializers.ModelSerializer):
     vouchers = serializers.SerializerMethodField()
     categories = CategorySerializer(many=True, read_only=True)
     note = serializers.SerializerMethodField()
+    usage_num = serializers.SerializerMethodField()
```
Comment (mjfrey) [inline ecommerce/extensions/api/serializers.py]: usage_count

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 3]
```diff
@@ -357,6 +375,32 @@ def test_exception_for_multi_use_voucher_type(self):
...
+        basket.add_product(product)
+        basket.vouchers.add(voucher)
+        Applicator().apply(basket, self.user)
+
+        # Verify that the voucher now has one possible application.
+        order = create_order(basket=basket, user=self.user)
+        lines = order.lines.all()
+        order, completed_lines = CouponFulfillmentModule().fulfill_product(order, lines)
+        self.assertEqual(completed_lines[0].status, LINE.COMPLETE)
+        self.assertEqual(voucher.offers.first().get_max_applications(), 1)
```
Comment (mjfrey) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Maybe add one more check to ensure the third time is rejected.

File: ecommerce/extensions/api/v2/views/coupons.py [hunk 4]
```diff
@@ -85,6 +85,10 @@ def create(self, request, *args, **kwargs):
             categories = Category.objects.filter(id__in=request.data['category_ids'])
             client, __ = BusinessClient.objects.get_or_create(name=client_username)
             note = request.data.get('note', None)
+            usage_num = request.data.get('usage_num', None)
+
+            if usage_num:
```
Comment (mjfrey) [inline ecommerce/extensions/api/v2/views/coupons.py]: usage _count is much better.

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 5]
```diff
@@ -253,6 +255,8 @@ def test_generate_coupon_report(self):
         self.assertEqual(field_names, [
             'Coupon Name',
             'Code',
+            'Maximum Coupon Limit',
```
Comment (mjfrey) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Maximum Coupon Usage

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 6]
```diff
@@ -75,12 +76,13 @@ def setup_coupons_for_report(self):
...
             catalog=self.catalog,
             coupon=self.coupon,
-            end_datetime=datetime.date(2015, 10, 30),
+            end_datetime=datetime.date(2099, 10, 30),
             name="Discount code",
             quantity=1,
             start_datetime=datetime.date(2015, 10, 1),
-            voucher_type=Voucher.SINGLE_USE,
-            code=VOUCHER_CODE
+            voucher_type=Voucher.ONCE_PER_CUSTOMER,
```
Comment (mjfrey) [inline ecommerce/extensions/voucher/tests/test_utils.py:92]: This type of voucher will cause issues when trying to display the offer to anonymous users.  Oscar checks that the voucher being used is valid for the user.  I suggest we move this to a MULTI-USE voucher type with max N usage.

Comment (vkaracic) [inline ecommerce/extensions/voucher/tests/test_utils.py:92]: That's a separate issue not introduced nor in the scope of this one since anonymous users have always gotten an error on the offer/redeem pages of `ONCE_PER_CUSTOMER` usage limitations.

I suggest filing a new ticket for this work.

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 5 above)
Comment (vkaracic) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Okay, changing `Limit` to `Usage` - FYI @pbaruah

File: ecommerce/extensions/api/serializers.py (hunk 1 above)
Comment (vkaracic) [inline ecommerce/extensions/api/serializers.py]: I'd rather rename it to `max_usage` if you agree.

Comment (vkaracic) [inline ecommerce/extensions/api/serializers.py]: I'd rather rename it to ``max_usage`` if you agree.

File: ecommerce/coupons/views.py [hunk 7]
```diff
@@ -85,15 +85,23 @@ def voucher_is_valid(voucher, product, request):
         elif voucher.end_datetime < now:  # pragma: no cover
             return False, _('This coupon code has expired.')
 
-    avail, msg = voucher.is_available_to_user(request.user)
-    if not avail:
-        voucher_msg = msg.replace('voucher', 'coupon')
-        return False, voucher_msg
+    # If the voucher is of type ONCE_PER_CUSTOMER we want to display the
+    # offer page to the anonymous user.
+    if request.user.is_authenticated():
```
Comment (mjfrey) [inline ecommerce/coupons/views.py:89]: I think this is true for all offers.  We want to show the offer to anonymous users.  We will check the validity of the offer once the user logs in.

File: ecommerce/coupons/views.py [hunk 8]
```diff
@@ -85,15 +85,23 @@ def voucher_is_valid(voucher, product, request):
...
+            return False, voucher_msg
 
     purchase_info = request.strategy.fetch_for_product(product)
     if not purchase_info.availability.is_available_to_buy:
         return False, _('Product [{product}] not available for purchase.'.format(product=product))
 
+    # If the voucher's number of applications exceeds it's limit.
+    offer = voucher.offers.first()
+    if offer.get_max_applications(request.user) == 0:
+        return False, _('The usage limit for this coupon has exceeded.')
```
Comment (mjfrey) [inline ecommerce/coupons/views.py]: Need better message here.  @pbaruah can you suggest?

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 9]
```diff
@@ -357,6 +375,40 @@ def test_exception_for_multi_use_voucher_type(self):
         with self.assertRaises(NotImplementedError):
             self.client.post(COUPONS_LINK, data=self.data, format='json')
 
+    def apply_voucher(self, voucher):
```
Comment (mjfrey) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: move this to CouponMixin

File: ecommerce/static/templates/coupon_form.html [hunk 10]
```diff
@@ -56,6 +56,12 @@
         </div>
 
         <div class="form-group">
+            <label for="max_usage"><%= gettext('Number of Code usages') %></label>
```
Comment (mjfrey) [inline ecommerce/static/templates/coupon_form.html]: Need better text here.  
"Max number of usage"

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=652

File: ecommerce/coupons/views.py (hunk 7 above)
Comment (vkaracic) [inline ecommerce/coupons/views.py:89]: You're suggesting to adjust the comment above that line?

File: ecommerce/settings/test.py [hunk 11]
```diff
@@ -124,5 +124,5 @@ def __getitem__(self, item):
 
 
 # Use production settings for asset compression so that asset compilation can be tested on the CI server.
-COMPRESS_ENABLED = True
-COMPRESS_OFFLINE = True
+COMPRESS_ENABLED = False
+COMPRESS_OFFLINE = False
```
Comment (mjfrey) [inline ecommerce/settings/test.py]: Remove this from the PR

File: ecommerce/coupons/views.py (hunk 7 above)
Comment (mjfrey) [inline ecommerce/coupons/views.py:89]: Yes. This change is for all Coupons not just ONCE_PER_CUSTOMER

File: ecommerce/extensions/api/serializers.py [hunk 12]
```diff
@@ -433,11 +434,19 @@ def get_note(self, obj):
         except AttributeError:
             return None
 
+    def get_max_usage(self, obj):
+        offer = obj.attr.coupon_vouchers.vouchers.first().offers.first()
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: This gets called multiple times. Time to make it a helper method.

File: ecommerce/extensions/api/serializers.py [hunk 13]
```diff
@@ -400,6 +400,7 @@ class CouponSerializer(ProductPaymentInfoMixin, serializers.ModelSerializer):
     vouchers = serializers.SerializerMethodField()
     categories = CategorySerializer(many=True, read_only=True)
     note = serializers.SerializerMethodField()
+    max_usage = serializers.SerializerMethodField()
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: I'm not a fan of unnecessary nesting. Why not have two separate fields—`num_uses` and `max_uses`?

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 14]
```diff
@@ -166,6 +171,19 @@ def test_coupon_note(self):
         self.assertEqual(note_coupon.attr.note, '𝑵𝑶𝑻𝑬')
         self.assertEqual(note_coupon.title, 'Coupon')
 
+    def test_multiple_usage_coupon(self):
+        """Test creating a multiple-usage coupon."""
+        self.coupon_data.update({
+            'max_usage': 2,
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Minor: Use a variable for `2`.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 15]
```diff
@@ -166,6 +171,19 @@ def test_coupon_note(self):
         self.assertEqual(note_coupon.attr.note, '𝑵𝑶𝑻𝑬')
         self.assertEqual(note_coupon.title, 'Coupon')
 
+    def test_multiple_usage_coupon(self):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: `test_multi_use_coupon_creation`

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 16]
```diff
@@ -166,6 +171,19 @@ def test_coupon_note(self):
...
+        """Test creating a multiple-usage coupon."""
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Verify the endpoint supports the creation of multi-usage coupons.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 17]
```diff
@@ -357,6 +375,31 @@ def test_exception_for_multi_use_voucher_type(self):
         with self.assertRaises(NotImplementedError):
             self.client.post(COUPONS_LINK, data=self.data, format='json')
 
+    def test_multiple_usage_coupon(self):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: This test doesn't belong here. It's not really an API test. It's a test of model logic, thus it belongs with the model tests.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 18]
```diff
@@ -357,6 +375,31 @@ def test_exception_for_multi_use_voucher_type(self):
...
+        """Test that multiple-usage coupon is created and the usage number decreased on usage."""
+        self.data.update({
+            'title': 'Multiple-usage',
+            'max_usage': 1,
+        })
+        # Verify that the created voucher has two possible applications.
+        response = self.client.post(COUPONS_LINK, data=self.data, format='json')
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Use a factory.

File: ecommerce/extensions/api/v2/views/coupons.py [hunk 19]
```diff
@@ -85,6 +85,10 @@ def create(self, request, *args, **kwargs):
             categories = Category.objects.filter(id__in=request.data['category_ids'])
             client, __ = BusinessClient.objects.get_or_create(name=client_username)
             note = request.data.get('note', None)
+            max_usage = request.data.get('max_usage', None)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/coupons.py]: `max_uses` seems more appropriate, and less confusing.

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 20]
```diff
@@ -273,39 +278,37 @@ def test_generate_coupon_report(self):
             'Coupon Expiry Date',
         ])
 
-        self.validate_report_of_redeemed_vouchers(rows[0], str(self.user.id), self.user.username)
-        self.validate_report_of_redeemed_vouchers(
-            rows[1],
-            '{}, {}'.format(self.user.id, user2.id),
-            '{}, {}'.format(self.user.username, user2.username)
+        discount_code_header_row = rows[1]
```
Comment (clintonb) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Hardcoded values like this are wrong. Write an assertion method (e.g. `assert_report_row(row, voucher)`) that can be used to compare the report data to a voucher. Iterate over the report rows and vouchers.

File: ecommerce/extensions/voucher/utils.py [hunk 21]
```diff
@@ -60,6 +134,7 @@ def generate_coupon_report(coupon_vouchers):
         _('Discount Percentage'),
         _('Discount Amount'),
         _('Status'),
+        _('Order Number'),
         _('Redeemed By ID'),
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Why is this field in the report? It's pretty worthless given that you have usernames.

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 22]
```diff
@@ -319,8 +322,34 @@ def test_generate_coupon_report(self):
...
+            catalog=self.catalog,
+            coupon=self.coupon,
+            end_datetime=datetime.date(2015, 10, 30),
+            name="Inactive code",
+            quantity=1,
+            start_datetime=datetime.date(2015, 10, 30),
+            voucher_type=Voucher.SINGLE_USE
+        )
+
+        self.mock_course_api_response(course=self.course)
```
Comment (clintonb) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Why is this necessary for a report? Digging into the code, you make an API call to get the value of org...which is innacurately labeld as "Partner". You can read the org locally from the course key/ID using the [`opaque_keys` library](https://pypi.python.org/pypi/edx-opaque-keys).

File: ecommerce/extensions/voucher/utils.py [hunk 23]
```diff
@@ -33,6 +33,78 @@
...
+    offer = voucher.offers.all().first()
+    stockrecords = offer.condition.range.catalog.stock_records.all()
+    course_seat = Product.objects.filter(id__in=[sr.product.id for sr in stockrecords]).first()
+    seat_stockrecord = course_seat.stockrecords.first()
+    history = coupon.history.latest()
+
+    coupon_stockrecord = StockRecord.objects.get(product=coupon)
+    course_id = course_seat.course.id
+    api = EdxRestApiClient(get_lms_url('api/courses/v1/'))
+    course_organization = api.courses(course_id).get()['org']
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Use `opaque_keys`.

File: ecommerce/extensions/voucher/utils.py [hunk 24]
```diff
@@ -33,6 +33,78 @@
 VoucherApplication = get_model('voucher', 'VoucherApplication')
 
 
+def _get_info_for_coupon_report(coupon, voucher):
+    offer = voucher.offers.all().first()
+    stockrecords = offer.condition.range.catalog.stock_records.all()
+    course_seat = Product.objects.filter(id__in=[sr.product.id for sr in stockrecords]).first()
+    seat_stockrecord = course_seat.stockrecords.first()
+    history = coupon.history.latest()
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: You want `first()` not `latest()`.

File: ecommerce/extensions/voucher/utils.py [hunk 25]
```diff
@@ -33,6 +33,78 @@
...
+
+    coupon_stockrecord = StockRecord.objects.get(product=coupon)
+    course_id = course_seat.course.id
+    api = EdxRestApiClient(get_lms_url('api/courses/v1/'))
+    course_organization = api.courses(course_id).get()['org']
+
+    price = currency(seat_stockrecord.price_excl_tax)
+    invoiced_amount = currency(coupon_stockrecord.price_excl_tax)
+    datetime_now = pytz.utc.localize(datetime.datetime.now())
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Nit: The word `localize` is a bit confusing. `datetime.datetime.now(pytz.UTC)` is clearer.

File: ecommerce/extensions/voucher/utils.py [hunk 26]
```diff
@@ -33,6 +33,78 @@
 VoucherApplication = get_model('voucher', 'VoucherApplication')
 
 
+def _get_info_for_coupon_report(coupon, voucher):
+    offer = voucher.offers.all().first()
+    stockrecords = offer.condition.range.catalog.stock_records.all()
+    course_seat = Product.objects.filter(id__in=[sr.product.id for sr in stockrecords]).first()
+    seat_stockrecord = course_seat.stockrecords.first()
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Isn't this just `offer.condition.range.catalog.stock_records.first()`? `course_seat` can also be derived from this value.

File: ecommerce/extensions/voucher/utils.py [hunk 27]
```diff
@@ -33,6 +33,78 @@
...
+    )
+    discount_data = get_voucher_discount_info(offer.benefit, seat_stockrecord.price_excl_tax)
+    coupon_type = _('Discount') if discount_data['is_discounted'] else _('Enrollment')
+    if in_datetime_interval:
+        status = _('Redeemed') if not offer.is_available() else _('Active')
+    else:
+        status = _('Inactive')
+
+    discount_percentage = _("{percentage} %").format(percentage=discount_data['discount_percentage'])
+    discount_amount = currency(discount_data['discount_percentage'])
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Typo?

File: ecommerce/extensions/voucher/utils.py [hunk 28]
```diff
@@ -33,6 +33,78 @@
...
+
+    coupon_stockrecord = StockRecord.objects.get(product=coupon)
+    course_id = course_seat.course.id
+    api = EdxRestApiClient(get_lms_url('api/courses/v1/'))
+    course_organization = api.courses(course_id).get()['org']
+
+    price = currency(seat_stockrecord.price_excl_tax)
+    invoiced_amount = currency(coupon_stockrecord.price_excl_tax)
+    datetime_now = pytz.utc.localize(datetime.datetime.now())
+    in_datetime_interval = (
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: not_expired

Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: This function is rather large. You can clean it up by breaking it into smaller functions. `get_voucher_status()` seems like a good function for what's going on here.

File: ecommerce/extensions/voucher/utils.py [hunk 29]
```diff
@@ -33,6 +33,78 @@
...
+    coupon_type = _('Discount') if discount_data['is_discounted'] else _('Enrollment')
+    if in_datetime_interval:
+        status = _('Redeemed') if not offer.is_available() else _('Active')
+    else:
+        status = _('Inactive')
+
+    discount_percentage = _("{percentage} %").format(percentage=discount_data['discount_percentage'])
+    discount_amount = currency(discount_data['discount_percentage'])
+
+    URL = '{url}?code={code}'.format(url=get_ecommerce_url('/coupons/offer/'), code=voucher.code)
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: No hard-coded paths. Use `reverse()`.

Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Also, `url`. This is not a constant.

File: ecommerce/extensions/voucher/utils.py [hunk 30]
```diff
@@ -33,6 +33,78 @@
...
+        'Partner': course_organization,
+        'Category': ProductCategory.objects.get(product=coupon).category.name,
+        'Note': note,
+        'Price': price,
+        'Invoiced Amount': invoiced_amount,
+        'Discount Percentage': discount_percentage,
+        'Discount Amount': discount_amount,
+        'Status': status,
+        'Created By': author,
+        'Create Date': history.history_date.strftime("%b %d,%y"),
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Consider using `.isoformat()`. Or, at least, at a space after the comma.

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 20 above)
Comment (clintonb) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Actually, if you validate `_get_info_for_coupon_report()` with tests, you can use it to validate the output of your endpoint. We do something similar for serializers.

File: ecommerce/extensions/voucher/utils.py [hunk 31]
```diff
@@ -82,78 +159,57 @@ def generate_coupon_report(coupon_vouchers):
...
+            info = _get_info_for_coupon_report(coupon, voucher)
+
+            row = {
+                'Coupon Name': info['Coupon Name'],
+                'Code': info['Code'],
+                'Maximum Coupon Usage': info['Maximum Coupon Usage'],
+                'Redemption Count': info['Redemption Count'],
+                'Coupon Type': info['Coupon Type'],
+                'URL': info['URL'],
+                'CourseID': info['CourseID'],
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Add a space: Course ID

File: ecommerce/extensions/voucher/utils.py [hunk 32]
```diff
@@ -82,78 +159,57 @@ def generate_coupon_report(coupon_vouchers):
...
-            discount_data = get_voucher_discount_info(offer.benefit, seat_stockrecord.price_excl_tax)
-            coupon_type = _('Discount') if discount_data['is_discounted'] else _('Enrollment')
-            if in_datetime_interval:
-                status = _('Redeemed') if voucher.num_orders > 0 else _('Active')
-            else:
-                status = _('Inactive')
-
-            discount_percentage = _("{percentage} %").format(percentage=discount_data['discount_percentage'])
-            discount_amount = currency(discount_data['discount_percentage'])
+            info = _get_info_for_coupon_report(coupon, voucher)
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]:
``` python
row = _get_info_for_coupon_report(coupon, voucher)

for item in ('Order Number', 'Redeemed By ID', 'Redeemed By Username',):
  row[item] = ''

rows.append(row)
```

File: ecommerce/static/templates/coupon_form.html [hunk 33]
```diff
@@ -56,6 +56,12 @@
         </div>
 
         <div class="form-group">
+            <label for="max_usage"><%= gettext('Maximum Number of Usage') %></label>
+            <input type="number" step="1" class="form-control" name="max_usage" value="1" min="1" max="10000">
```
Comment (clintonb) [inline ecommerce/static/templates/coupon_form.html]: Technically the max is 2147483647 for a [`PositiveIntegerField`](https://docs.djangoproject.com/en/1.9/ref/models/fields/#positiveintegerfield). Remove the `max` attribute just in case someone actually needs a ridiculous number of usages.

File: ecommerce/static/templates/coupon_form.html [hunk 34]
```diff
@@ -56,6 +56,12 @@
         </div>
 
         <div class="form-group">
+            <label for="max_usage"><%= gettext('Maximum Number of Usage') %></label>
```
Comment (clintonb) [inline ecommerce/static/templates/coupon_form.html]: s/Usage/Uses

File: ecommerce/static/js/views/coupon_form_view.js [hunk 35]
```diff
@@ -202,6 +205,13 @@ define([
                     formGroup('[name=code]').addClass('hidden');
                 }
 
+                // When creating a Once by multiple customers code show the usage number field.
+                if (voucherType !== 'Single use') {
+                    formGroup('[name=max_usage]').removeClass('hidden');
```
Comment (clintonb) [inline ecommerce/static/js/views/coupon_form_view.js]: `'hidden'` should be a constant.

File: ecommerce/extensions/api/serializers.py (hunk 13 above)
Comment (vkaracic) [inline ecommerce/extensions/api/serializers.py]: That would result in an additional database hit. In fact this whole serializer will probably be refactored to optimize for speed in the next sprint.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py (hunk 17 above)
Comment (vkaracic) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: I'll add a test for the model logic, but I'd like to keep this one as this tests that `create_coupon_product` supports multi-usage coupons.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py (hunk 18 above)
Comment (vkaracic) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Please elaborate.

File: ecommerce/extensions/voucher/utils.py (hunk 21 above)
Comment (vkaracic) [inline ecommerce/extensions/voucher/utils.py]: Should I remove this field, @pbaruah ?

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 22 above)
Comment (vkaracic) [inline ecommerce/extensions/voucher/tests/test_utils.py]: I think this is wrong, what we want is probably the site partner(`stockrecord.partner.name`) not the course org. Right, @mjfrey?

File: ecommerce/extensions/voucher/utils.py (hunk 30 above)
Comment (vkaracic) [inline ecommerce/extensions/voucher/utils.py]: I think this is the format that product wants for the report. 
@pbaruah what kind of date format goes here? Currently it's:
`Apr 12,2016`
isoformat would be:
`2016-04-12`

File: ecommerce/static/templates/coupon_form.html (hunk 33 above)
Comment (vkaracic) [inline ecommerce/static/templates/coupon_form.html]: @mjfrey and @pbaruah let's discuss this part
(in my offence the max value stated in the ticket is 100000 not 10000)

File: ecommerce/static/js/views/coupon_form_view.js (hunk 35 above)
Comment (vkaracic) [inline ecommerce/static/js/views/coupon_form_view.js]: Please elaborate.

File: ecommerce/extensions/voucher/tests/test_utils.py (hunk 22 above)
Comment (mjfrey) [inline ecommerce/extensions/voucher/tests/test_utils.py]: We want the course organization.  Use the opaque_keys to obtain and change the Heading of the report to Org not Partner.

Comment (vkaracic) [inline ecommerce/extensions/voucher/tests/test_utils.py]: opaque_keys library needs to be installed first, can I add it to the requirements?

Comment (clintonb) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Yes.

File: ecommerce/extensions/voucher/utils.py (hunk 21 above)
Comment (mjfrey) [inline ecommerce/extensions/voucher/utils.py]: Yes remove the ID as we have the username

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py (hunk 17 above)
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: You're already testing coupon creation in `test_multi_use_coupon_creation` above.

File: ecommerce/coupons/views.py (hunk 8 above)
Comment (clintonb) [inline ecommerce/coupons/views.py]: @srpearce ?

**Comment by clintonb** [comment]
:+1: once remaining comments addressed.

File: ecommerce/coupons/views.py (hunk 8 above)
Comment (pbaruah) [inline ecommerce/coupons/views.py]: @vkaracic - this is the message @srpearce suggested - 

"This coupon code is no longer available"
